File: csharp/DVSClient/DVSClient/address/format/FormatResultFuture.cs
using DVSClient.Exceptions;
using DVSClient.Server.Address.Format;

namespace DVSClient.Address.Format
{
    public class FormatResultFuture
    {
        private readonly Task<RestApiAddressFormatResponse> _apiFuture;

        public FormatResultFuture(Task<RestApiAddressFormatResponse> apiFuture)
        {
            _apiFuture = apiFuture;
        }

        public async Task<FormatResult> GetAsync()
        {
            try
            {
                var response = await _apiFuture;
                if (response.Error != null)
                {
                    throw EDVSException.Using(response.Error);
                }
                return new FormatResult(response);
            }
            catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException)
            {
                throw new RestApiInterruptionOrExecutionException(e);
            }
        }

        public async Task<FormatResult> GetAsync(TimeSpan timeout)
        {
            try
            {
                var response = await Task.WhenAny(_apiFuture, Task.Delay(timeout)) == _apiFuture
                    ? await _apiFuture
                    : throw new TimeoutException();

                if (response.Error != null)
                {
                    throw EDVSException.Using(response.Error);
                }
                return new FormatResult(response);
            }
            catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException || e is TimeoutException)
            {
                throw new RestApiInterruptionOrExecutionException(e);
            }
        }

        public bool Cancel(bool mayInterruptIfRunning)
        {
            return _apiFuture.IsCanceled;
        }

        public bool IsCancelled()
        {
            return _apiFuture.IsCanceled;
        }

        public bool IsDone()
        {
            return _apiFuture.IsCompleted;
        }
    }
}

File: java/src/main/java/com/experian/dvs/client/address/format/FormatResultFuture.java
package com.experian.dvs.client.address.format;

import com.experian.dvs.client.exceptions.EDVSException;
import com.experian.dvs.client.server.address.format.RestApiAddressFormatResponse;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class FormatResultFuture implements Future<FormatResult> {

    private final Future<RestApiAddressFormatResponse> apiFuture;

    public FormatResultFuture(Future<RestApiAddressFormatResponse> apiFuture) {
        this.apiFuture = apiFuture;
    }

    @Override
    public boolean cancel(boolean mayInterruptIfRunning) {
        return this.apiFuture.cancel(mayInterruptIfRunning);
    }

    @Override
    public boolean isCancelled() {
        return this.apiFuture.isCancelled();
    }

    @Override
    public boolean isDone() {
        return this.apiFuture.isDone();
    }

    @Override
    public FormatResult get() throws InterruptedException, ExecutionException {
        final RestApiAddressFormatResponse response = this.apiFuture.get();
        if (response.getError() != null) {
            throw EDVSException.using(response.getError());
        }
        return new FormatResult(response);
    }

    @Override
    public FormatResult get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
        final RestApiAddressFormatResponse response = this.apiFuture.get(timeout, unit);
        if (response.getError() != null) {
            throw EDVSException.using(response.getError());
        }
        return new FormatResult(response);
    }
}
